[RPCSAGA] rescanblockchain core compatible. (#558)
getfloresta/Floresta: doc/rpc/rescanblockchain.md

# `rescanblockchain`

Sends a request to the node for rescan the blockchain searching for addresses and utxos.

## Arguments

> * start_height (numeric, optional, default=0): The initial block height to start the blockchain rescan. When timestamp is specified this numeric value will be understood as a block UNIX timestamp.

Reviewer: ```suggestion
* start_height (numeric, optional, default=0): The initial block height to start the blockchain rescan. When timestamp is set this numeric value will be understood as a block's UNIX timestamp.
```

> * start_height (numeric, optional): The initial height to start rescanning. When timestamp is specified this numeric value will be understood as a block UNIX timestamp.
> 
> * stop_height (numeric, optional): The height limit to stop the rescanning. When timestamp is specified this numeric value will be understood as a block UNIX timestamp.
> 
> * use_timestamp (boolean flag, optional): When present in the command the inserted values will be treated as unix timestamps.

Reviewer 2: In the code it says the default value for stop height is 0, but it would not make much sense putting it here. 

```suggestion
* start_height (numeric, optional, default = 0): The initial block height to start the blockchain rescan. When timestamp is specified this numeric value will be understood as a block UNIX timestamp.

* stop_height (numeric, optional): The height limit to stop the rescanning. When timestamp is specified this numeric value will be understood as a block UNIX timestamp.

* use_timestamp (boolean flag, optional): When present in the command the inserted values will be treated as UNIX timestamps.
```

> * use_timestamp (boolean flag, optional): When present in the command the inserted values will be treated as UNIX timestamps.

Reviewer: nit: `s/inserted/provided/`

## Returns

### Ok Response

When the rescan request sucessfully starts you will receive an Ok(true).

### Error Enum [`CommandError`]

This RPC command doesnt have any specific error enum besides the internal calls related to blockchain requests.

## Usage Examples

```bash

# Rescan from height 100 to 200

floresta-cli rescanblockchain 100 200

# Rescan from height 100 to tip

floresta-cli rescanblockchain 100

# Rescan from timestamp 1231006505 (genesis) until 133456789

floresta-cli rescanblockchain -t 1231006505 1752516460

# Rescan from timestamp 0 (alias for genesis) until the tip

floresta-cli rescanblockchain --timestamp 0

 ```

## Notes

> - Be sure to not insert invalid values, e.g. the start being greater than the stop.

Reviewer: Can't we check for this and simply return earlier?

# `rescanblockchain`

Sends a request to the node for rescan the blockchain searching for transactions related to the wallets cached addresses.

## Arguments

> * start_height (numeric, optional, default=0): The initial block height to start the blockchain rescan.

Reviewer 3: ```suggestion
* `start_height` (numeric, optional, default=`0`): The initial block height to start the blockchain rescan.
```

# `rescanblockchain`

> Sends a request to the node for rescan the blockchain searching for transactions related to the wallets cached addresses.

Reviewer: Here too wallet's

* `stop_height` (numeric, optional, default=`0`): The height limit to stop the rescanning.

* `use_timestamp` (boolean flag, optional, default=`false`): When present in the command the provided values will be treated as UNIX timestamps. These timestamps do not need to be directly from a block and can be used for finding addresses and UTXOs from meaningful timestamp values.

* `confidence` (string, optional, default=`medium`): In cases where `use_timestamp` is set, tells how much confidence the user wants for finding its addresses from this rescan request, a higher confidence will add more lookback seconds to the targeted timestamp and rescanning more blocks.

Under the hood this uses an [Exponential distribution](https://en.wikipedia.org/wiki/Exponential_distribution) [cumulative distribution function (CDF)](https://en.wikipedia.org/wiki/Cumulative_distribution_function) where the parameter $\lambda$ (rate) is $\frac{1}{600}$ (1 block every 600 seconds, 10 minutes).
  The supplied string can be one of:
  
  - `high`: 99% confidence interval.
  - `medium` (default): 95% confidence interval.
  - `low`: 90% confidence interval.

>   - `none`: Doesnt apply any lookback seconds.

Reviewer: Need to update the variant name here

## Returns

### Ok Response

> When the rescan request sucessfully starts you will receive an Ok(true).

Reviewer: After the rescan finishes, you'll see a log message telling all the blocks that may have any transactions to your wallet

### Error Enum

This RPC command can only fail if:
   - we can't communicate with the headers database;
   - if invalid values are provided. That is, the start of the request being less than the stop value.

> If `timestamp` is true, apart the previous rules, if any of the values is smaller than that of the genesis block (1231006505) or greater than the tip time, the execution will also fail.

Reviewer: Smaller than the network's genesis block time (1231006505 for mainnet)

Furthermore, the request will be aborted if the node still syncing with the blockchain.

## Usage Examples

```bash

# Rescan from height 100 to 200

floresta-cli rescanblockchain 100 200

# Rescan from height 100 to tip

floresta-cli rescanblockchain 100

# Rescan from timestamp 1231006505 (genesis) until 133456789

floresta-cli rescanblockchain -t 1231006505 1752516460 --confidence high

# Rescan from timestamp 0 (alias for genesis) until the tip

floresta-cli rescanblockchain --timestamp 0 1752516460 -c high

 ```

## Notes

- Be sure to not insert invalid values, e.g. the start being greater than the stop.

- This rescan relies on BIP 158 block filters.

> - You dont need to be picky with timestamps but, when using uncertain timestamps you mostly want to set a high confidence which is not necessary for precise timestamps.

Reviewer 2: Shouldn't it be a "lower confidence"? As a higher confidence means a `p = 0.01` and a lower confidence means `p = 0.10`

Author: Nuh uh
